File: go/src/panagram.go
package main

import "os"

type OptionType string 

const (
	Palindrome OptionType = "PALINDROME"
	Anagram = "ANAGRAM"
	Help = "HELP"
	Quit = "QUIT"
	Unknown = ""
)

func toOptionType(mode string) OptionType {
	switch mode {
		case "p": return Palindrome;
		case "a": return Anagram;
		case "h": return Help;
		case "q": return Quit;
		default: return Unknown;
	}
}

type Panagram struct {
	consoleIO ConsoleIO
}

func NewPanagram() *Panagram {
	return &Panagram{}
}

func (p *Panagram) BadArgs(tooMany bool, option string) {
	msg := "Too "
	if tooMany {
		msg += "many"
	} else {
		msg += "few"
	}
	msg += " arguments for option " + option
	p.consoleIO.WriteStderr(msg)
}

func (p *Panagram) BadMode(mode string) {
	p.consoleIO.WriteStderr("Unknown option " + mode)
}

func isOrNot(ifTrue bool) string {
	txt := "is"
	if !ifTrue {
		txt += " not"
	}
	return txt
}

func (p *Panagram) StaticMode() {
	var argCount = len(os.Args)

	// CAUTION this will not work for mult-byte characters
	var modeArg = os.Args[1]
	var mode = modeArg[1:]

	var option = toOptionType(mode);

	switch option {
	case Anagram:
		if (argCount != 4) {
			p.BadArgs(argCount > 4, mode)
		} else {
			var arg1 = os.Args[2]
			var arg2 = os.Args[3]
			p.consoleIO.WriteStdout(arg2 + " " + isOrNot(AreAnagrams(arg1, arg2)) + " an anargram of " + arg1)
		}
		break;
	case Palindrome:
		if argCount != 3 {
			p.BadArgs(argCount > 3, mode)
		} else {
			var arg = os.Args[2]
			p.consoleIO.WriteStdout(arg + " " + isOrNot(IsPalindrome(arg)) + " a palindrome")
		}
		break;
	case Help:
		p.consoleIO.PrintUsage()
		break;
	default:
		p.BadMode(mode)
		p.consoleIO.PrintUsage()
	}

	// p.consoleIO.WriteStdout("You supplied mode " + mode)
}

func (p *Panagram) InteractiveMode() {
	p.consoleIO.WriteStdout("Welcome to Panagram. This program checks if an input string is an anagram or palindrome.")

	var shouldQuit = false;
	for; !shouldQuit; {
		p.consoleIO.WriteStdout("Type 'a' to check for anagrams or 'p' for palindromes type 'q' to quit.")

		var input = p.consoleIO.GetInput()
		var mode = toOptionType(input)

		switch mode {
		case Anagram:
			p.consoleIO.WriteStdout("Type the first string:");
			var arg1 = p.consoleIO.GetInput();
			p.consoleIO.WriteStdout("Type the second string:");
			var arg2 = p.consoleIO.GetInput();

			p.consoleIO.WriteStdout(
					arg2 + " " + isOrNot(AreAnagrams(arg1, arg2)) + " an anagram of " + arg1)
			break;
			case Palindrome:
				p.consoleIO.WriteStdout("Type a word or sentence:")
				var arg = p.consoleIO.GetInput()
				p.consoleIO.WriteStdout(arg + " " + isOrNot(IsPalindrome(arg)) + " a palindrome")
				break;
			case Quit:
				shouldQuit = true
				break;
			default:
				p.BadMode(input)
		}
	} 

}

File: dart/lib/Panagram.dart
import 'ConsoleIO.dart';
import 'String_extensions.dart';

enum OptionType { palindrome, anagram, help, unknown, quit }

OptionType toOptionType(String mode) {
  switch (mode) {
    case 'p':
      return OptionType.palindrome;
    case 'a':
      return OptionType.anagram;
    case 'h':
      return OptionType.help;
    case 'q':
      return OptionType.quit;
    default:
      return OptionType.unknown;
  }
}

class Panagram {
  final consoleIO = ConsoleIO();

  void badArgs(bool tooMany, String option) {
    consoleIO.write(
        "Too ${tooMany ? "many" : "few"} arguments for option $option",
        to: OutputType.error);
    consoleIO.printUsage();
  }

  void badMode(String mode) {
    consoleIO.write('Unknown option $mode', to: OutputType.error);
  }

  String isOrNot(bool ifTrue) {
    return "is${ifTrue ? "" : " not"}";
  }

  void staticMode(List<String> args) {
    final argCount = args.length;
    final modeArg = argCount > 0 ? args.first : '-';

    final value = modeArg.substring(1);

    var option = toOptionType(value);

    switch (option) {
      case OptionType.anagram:
        if (argCount != 3) {
          badArgs(argCount > 3, value);
        } else {
          final arg1 = args[1];
          final arg2 = args[2];
          consoleIO.write(
              '$arg2 ${isOrNot(arg1.isAnagramOf(arg2))} an anagram of $arg1');
        }
        break;

      case OptionType.palindrome:
        if (argCount != 2) {
          badArgs(argCount > 3, value);
        } else {
          final arg = args[1];
          consoleIO.write('$arg ${isOrNot(arg.isPalindrome())} a palindrome');
        }
        break;

      case OptionType.help:
        consoleIO.printUsage();
        break;

      default:
        badMode(value);
        consoleIO.printUsage();
    }
  }

  void interactiveMode() {
    consoleIO.write(
        'Welcome to Panagram. This program checks if an input string is an anagram or palindrome.');

    var shouldQuit = false;
    while (!shouldQuit) {
      consoleIO.write(
          "Type 'a' to check for anagrams or 'p' for palindromes type 'q' to quit.");

      final mode = getOption(consoleIO.getInput());
      final OptionType option = mode[0];
      final String value = mode[1];

      switch (option) {
        case OptionType.anagram:
          consoleIO.write('Type the first string:');
          final arg1 = consoleIO.getInput();
          consoleIO.write('Type the second string:');
          final arg2 = consoleIO.getInput();

          consoleIO.write(
              '$arg2 ${isOrNot(arg1.isAnagramOf(arg2))} an anagram of $arg1');
          break;

        case OptionType.palindrome:
          consoleIO.write('Type a word or sentence:');
          final arg = consoleIO.getInput();
          consoleIO.write('$arg ${isOrNot(arg.isPalindrome())} a palindrome');
          break;

        case OptionType.quit:
          shouldQuit = true;
          break;

        default:
          badMode(value);
      }
    }
  }

  List getOption(String option) {
    return [toOptionType(option), option];
  }

}
